Handle implicit member references inside extensions of nested types

jrose-apple opened the pull request:
Within an extension, references to other members of the extended type are permitted without qualification. This is intended to work even when the extended type was a nested type, although members of the enclosing type are *not* visible in this case. In order to implement this, the type checker pre-checks to see if there are *any* members with this name and then rewrites the unqualified reference to a qualified one, based on an unresolved TypeExpr with the name of the enclosing type. Unfortunately, if the enclosing type is a nested type, that isn't going to work very well—we find the correct declaration, but fail to map it into context by virtue of not realizing where it came from. Fix this by explicitly checking for this case.

[SR-3847](https://bugs.swift.org/browse/SR-3847)

jrose-apple commented:
@swift-ci Please test

slavapestov commented:
Oh I see, thanks.

jrose-apple commented:
Does that mean the fix makes sense? I admit I'm not very familiar with this part of name resolution.

jrose-apple commented:
@swift-ci Please test

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -264,6 +264,11 @@ findDeclContextForType(TypeChecker &TC,
       return std::make_tuple(parentDC, parentNominal, true);
 
+    if (isa<ExtensionDecl>(parentDC) && typeDecl == parentNominal) {
+      assert(parentDC->getParent()->isModuleScopeContext());
```
An extension can appear inside another type, though. We diagnose it as invalid but still attempt to do name lookup into it.

jrose-apple replied:
Hm. Guess I'd better add an invalid test case for that.

slavapestov commented:
mapTypeIntoContext() translates interface types to archetypes. Here, I think the code path is resolving an unbound generic reference, which implicitly gets generic parameters from context.

slavapestov commented:
LGTM, by the way. Longer term, we should try to remove findDeclContextForType() altogether. It duplicates the DeclContext walk that's already done in AST/NameLookup.cpp, because we just throw out that information in UnqualifiedLookup. That's really stupid and it causes assertions because apparently the walk here is slightly different than the one name lookup does...

jrose-apple committed:
Handle implicit member references inside extensions of nested types.

Within an extension, references to other members of the extended type
are permitted without qualification. This is intended to work even
when the extended type was a nested type, although members of the
enclosing type are /not/ visible in this case. In order to implement
this, the type checker pre-checks to see if there are /any/ members
with this name and then rewrites the unqualified reference to a
qualified one, based on an unresolved TypeExpr with the name of the
enclosing type. Unfortunately, if the enclosing type is a nested type,
that isn't going to work very well---we find the correct declaration,
but fail to map it into context by virtue of not realizing where it
came from. Fix this by explicitly checking for this case.

https://bugs.swift.org/browse/SR-3847

jrose-apple commented:
Test case added.

@swift-ci Please smoke test

jrose-apple commented:
[Unrelated crash in serialization](https://ci.swift.org/job/swift-PR-Linux-smoke-test/4502/). @aciidb0mb3r, have you seen this before? Did you file it already?

```
3.	While type-checking expression at [/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swiftpm/Sources/PackageGraph/PackageGraphLoader.swift:111:35 - line:111:118] RangeText="externalModuleDependencies.lazy.map({$0.name}).duplicates(modules.lazy.map{$0.name})"
4.	While loading members for 'LazyMapRandomAccessCollection' in module 'Swift'
5.	While deserializing 'init' (ConstructorDecl #11643)in 'Swift'
6.	While deserializing 'transform' (ParamDecl #22123)in 'Swift'
```

@swift-ci Please smoke test Linux

aciidb0mb3r commented:
@jrose-apple Haven't seen that one, seems new

jrose-apple commented:
[*Another* failure](https://ci.swift.org/job/swift-PR-Linux-smoke-test/4506).

> swift: /home/buildnode/disk2/workspace/swift-PR-Linux-smoke-test/branch-master/swift/lib/AST/ArchetypeBuilder.cpp:2484: auto swift::ArchetypeBuilder::getGenericEnvironment(swift::GenericSignature *)::(anonymous class)::operator()(swift::ArchetypeBuilder::PotentialArchetype *) const: Assertion `(inContext->isEqual(repInContext) || inContext->hasError() || repInContext->hasError()) && "Potential archetype mapping differs from representative!"' failed.

@swift-ci Please smoke test Linux
